### DISCUSSION TITLE
`loop` statement equivalent to `while (true)`

### DISCUSSION THREAD

**Discussion by Rekkonnect** [dotnet/csharplang#6643] [Language Ideas]
C# currently only offers loops that are bound to specific operations, `while` which breaks once the condition isn't met, `for` which provides a handy initialization and incrementation syntax sugar, and `foreach` which iterates over collections.

I've found myself using `while (true)` almost exclusively in certain scenarios, because most of the times the conditions to break are complex and would only be more problematic if I had to integrate it inside the condition. Many are the times that the breaking condition is in the middle of the looped body, meaning your best option is to evaluate the condition in its place and manually break the loop.

Rust provides a `loop` keyword and supports a native construct for `while (true)` and would feel more intuitive to type, as the developer is then asked to manually break the loop inside the body on demand without accounting for one special condition that's declared in a point above.

**Comment by Richiban** [comment]
The `loop` keyword in Rust isn't there just because it's shorter, it's there because certain features require knowing that the code is executed at least once (definite assignment analysis, etc) so there would have to be the same advantages in C# to make such a feature worthwhile. Simply having a syntactic shortcut for `while (true)` isn't going to cut it, I don't think.

**Reply by FaustVX** [reply]
So the `loop` loop in Rust is like the `do ... while()` loop in C# (but without the condition) ?

**Reply by Rekkonnect** [reply]
`while (true)` is the same as a `do while (true)`

**Reply by FaustVX** [reply]
@Rekkonnect 
Not exactly,
The second one means the loop will be executed at least once.

**Reply by sharwell** [reply]
`while (true)` also means the loop will be executed at least one.

**Reply by FaustVX** [reply]
Ok, I'm so dumb,
In my head, I thought of `while (cond)` vs `do … while (cond)`
Sorry for that.

**Reply by The-Futurist** [reply]
> The `loop` keyword in Rust isn't there just because it's shorter, it's there because certain features require knowing that the code is executed at least once (definite assignment analysis, etc) so there would have to be the same advantages in C# to make such a feature worthwhile. Simply having a syntactic shortcut for `while (true)` isn't going to cut it, I don't think.

The [Ada language](https://en.wikibooks.org/wiki/Ada_Programming/Control#Loops) too has a `loop` keyword, used alone it represents a true, infinite loop.

**Comment by jnm2** [comment]
See also https://github.com/dotnet/csharplang/issues/2475 for conditionless `while`

**Comment by FaustVX** [comment]
You can use the `while` loop with a condition
```C#
var myComplexCondition = true;
while (myComplexCondition)
{
    …
    if (…)
        myComplexCondition = false;
    …
}
```

**Reply by Rekkonnect** [reply]
My original proposal already addresses that situation and its whole purpose is to simplify this workaround down to creating a loop construct that solely relies on the developer manually breaking it inside its body.

**Reply by jnm2** [reply]
@FaustVX It's worth pointing out that this behaves differently than breaking in the middle of the loop, which is what this proposal was comparing to.

**Reply by FaustVX** [reply]
@jnm2 Yes, I didn't think of that.
I rarely `break` in the middle of a loop.

**Comment by TonyValenti** [comment]
Doesn't this do what you want?

```
for (;;) { 
   //body here            
}
```

**Reply by Rekkonnect** [reply]
`for (;;)` is harder to write than `whilte (true)` so why even resort to that? Plus, that statement is syntactical abuse, it was not quite intended to be used like that. For the same reason why you'd write `while (i < 5)` instead of `for (; i < 5;)`; `for` derived from `while`, not vice versa.

The point is to introduce a single keyword to replace an expression that consists of 4 tokens (`while`, `(`, `true`, `)`), not introduce even more clutter to the expression with `for (;;)` which bears 5 explicit (`for`, `(`, `;`, `;`, `)`) and 3 implicit tokens (between `(;`, between `;;`, between `;)`).

**Reply by CyrusNajmabadi** [reply]
> it was not quite intended to be used like that. 

It definitely was intended for use in that way. The spec makes it clear that all parts are optional, and what the semantics are for when a part is or is not there: https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/statements.md#1294-the-for-statement

This loop syntax and semantics were taken from our C lineage, where this behavior was known and intended as well. It was quite intentional for it to work this way. 

You may want a syntax with less tokens. That's certainly a totally reasonable request to make. But it's not the case that existing idioms were not intended for his purpose :-)

**Reply by Rekkonnect** [reply]
> It definitely was intended for use in that way.

In the C# world, yes. It would otherwise cause compilation errors. In the C world though, you had `while` for a standard conditional loop, and `for` for a loop that initialized and incremented values per iteration. You had the option to (ab)use `for (;;)`, but it was avoided in favor of `while (1)`.

Now, allowing `for (;;)` in the first place in C is a design decision, which should not happen if you wanted your language to not be "hacky" (bodge things till it works and make random exotic stuff work). C# took that initial design, no biggie. It's up to us to not use it, when we have `while (true)` which is easier to type and more quickly understandable.

**Reply by CyrusNajmabadi** [reply]
> which should not happen if you wanted your language to not be "hacky" (bodge things till it works and make random exotic stuff work)

It's neither random, nor exotic. Every part of this construct is intentionally optional, with clear and sensible semantics about what it means when not provided. There are no hacks here. 

You may personally find using such a construct undesirable, but that's a personal preference decision. Importantly, we don't use an individual's distastes to motivate adding redundant syntax.

**Reply by CyrusNajmabadi** [reply]
> which is easier to type and more quickly understandable.

Both are idiomatic C#. The language has been in around and in development for around 25 years now. Replacing idiomatic things present in v1 is not a goal.

**Reply by The-Futurist** [reply]
> Doesn't this do what you want?
> 
> ```
> for (;;) { 
>    //body here            
> }
> ```

True, @TonyValenti yet so does this:

```cs
repeat: 
   // body statement 1
   // body statement 2
goto repeat;
```

Inelegant though that it is, it doesn't even need braces and is somewhat _more_ elegant than a `for (;;)` an utterly cryptic, bizarre way to express unbounded repetition, they could have at least allowed `for()` for such a case or even `for ever`, I mean we already have `yield return` so surely this would be trivial.

**Comment by The-Futurist** [comment]
> C# currently only offers loops that are bound to specific operations, `while` which breaks once the condition isn't met, `for` which provides a handy initialization and incrementation syntax sugar, and `foreach` which iterates over collections.
> 
> I've found myself using `while (true)` almost exclusively in certain scenarios, because most of the times the conditions to break are complex and would only be more problematic if I had to integrate it inside the condition. Many are the times that the breaking condition is in the middle of the looped body, meaning your best option is to evaluate the condition in its place and manually break the loop.
> 
> Rust provides a `loop` keyword and supports a native construct for `while (true)` and would feel more intuitive to type, as the developer is then asked to manually break the loop inside the body on demand without accounting for one special condition that's declared in a point above.

It would be helpful to support a simple loop directive where the loop termination (if any) is handled by logic wholly in the loop body.

The reason we use contrivances like:

```cs
while (true)
```

or 

```cs
for (;;)
```

Is because the language does not offer an unconditional looping structure/keyword.

If adding a new keyword to C#'s grammar was trivial we could just consider adding a new "loop" or "forever" keyword, it would be that simple, done, dusted:

```cs
loop
{
   a = do_stuff();

   if (a == b)
      break;
}
```

I've been designing an unrelated language grammar that has this, all loops are variants of "loop" with optional clauses for "while" and "until" and iterations and so on. A "loop" is effectively always infinite, unconditional, _unless_ one includes additional clauses.

I don't think a suggestion for supporting a new simple way to express unconditional loops in C# is likely to go anywhere myself.

**Reply by spydacarnage** [reply]
> If adding a new keyword to C#'s grammar was trivial we could just consider adding a new "loop" or "forever" keyword, it would be that simple, done, dusted:

As has been discussed in other threads, adding a new keyword is _not_ trivial.

> I don't think a suggestion for supporting a new simple way to express unconditional loops in C# is likely to go anywhere myself.

Based on the above non-trivial nature of adding the keyword, versus the benefit of being able to type `loop` instead of `while (true)`, I think you are correct - I see no real benefit of this moving forward.

You could even write an analyzer to make sure there is a `break` statement inside a `while (true)` loop if you want to ensure that.

**Reply by CyrusNajmabadi** [reply]
> If adding a new keyword to C#'s grammar was trivial 

From a technical standpoint, adding a new keyword is trivial.  The downside is the complexity gain to the language/community/ecosystem.  In order to take on that sort of change, we need to believe that the benefit will be worth it there.

> Is because the language does not offer an unconditional looping structure/keyword.

Correct.  The language (and designers) does not think it needs to offer a targeted solution in-lang for all situations.  If we believe a scenario is handled just fine by existing solutions, then that will often be fine with us.  In other words, we think it's better for the language to have a smaller set of features, which can be used for a broader set of scenarios, even if that means that some scenarios don't have an absolute 'minimal' solution for that particular need.

> I don't think a suggestion for supporting a new simple way to express unconditional loops in C# is likely to go anywhere myself.

Feature requests are assessed based on the value they bring, against the downsides they incur.  This very much includes not wanting the lang to feel like a kitchen sink where everything has been thrown in.  Given that our back-compat guarantees are *extremely* high, this means that we would certainly have all of the same keywords as before, and now would have this one as well.  This would make things more complex, while not really improving things substantively for users.

**Reply by HaloFour** [reply]
I also firmly believe that loops without a condition should be slightly more painful to write so that the developer properly considers what they intend to do.

**Comment by The-Futurist** [comment]
Yes, we've discussed the subject of keywords and reserved words and so on before, it's deeply rooted in grammar design. If C# had an easily extensible grammar some of these suggestions would be trivial to deliver, no risk, little effort. But it has a grammar based on C and that decisions carries costs/benefits, this is one of the prices you pay for that decision.

**Reply by CyrusNajmabadi** [reply]
@Korporal as mentioned in prior conversations (and even mentioned literally directly above).  This is not a matter of grammar design.  Extending C# to support this is trivial.  The question is whether extending the language in this way is a *good* idea or not.  Currently, the philosophy of C# is that limiting such extensions and making extremely judicious choices where we do extend is the best thing for our ecosystem.  I've marked your post off-topic as it is not actually about this suggestion, and seems to be a meta-point about something unrelated.   Please keep things on-topic.  Here, that would be about if such an extension has enough value to outweigh the problems we think may also some with it.  

Note: this applies to other discussions as well.  Popping into discussions to lament about how you think the grammar should have been designed isn't going to fly here.  This was previously contained to https://github.com/dotnet/csharplang/discussions/5311, but didn't seem to be constructive, and it was locked.  That's not a pass to then leak the same stuff out into unrelated discussions about other future language features.

I think we would allow a dedicated discussion thread by you on hte topic of the C# grammar and the design decisions there.  Such questions and opinions would be fine in the context of that discussion, if we can keep it primarily on C# and not have it bleed out elsewhere.  Feel free to create such a discussion if you would like.  It would also be fine to bring these off-topic conversations to Discord (discord.gg/csharp) to discuss with the team and community there.  Discord is much more free-form, where all sorts of conversations and topic flows are totally acceptable.  CSharplang discussions need to stay on topic though.  Thanks!

**Comment by The-Futurist** [comment]
> @Korporal as mentioned in prior conversations (and even mentioned literally directly above). This is not a matter of grammar design. Extending C# to support this is trivial. The question is whether extending the language in this way is a _good_ idea or not. Currently, the philosophy of C# is that limiting such extensions and making extremely judicious choices where we do extend is the best thing for our ecosystem. I've marked your post off-topic as it is not actually about this suggestion, and seems to be a meta-point about something unrelated. Please keep things on-topic. Here, that would be about if such an extension has enough value to outweigh the problems we think may also some with it.
> 
> Note: this applies to other discussions as well. Popping into discussions to lament about how you think the grammar should have been designed isn't going to fly here. This was previously contained to #5311, but didn't seem to be constructive, and it was locked. That's not a pass to then leak the same stuff out into unrelated discussions about other future language features.
> 
> I think we would allow a dedicated discussion thread by you on hte topic of the C# grammar and the design decisions there. Such questions and opinions would be fine in the context of that discussion, if we can keep it primarily on C# and not have it bleed out elsewhere. Feel free to create such a discussion if you would like. It would also be fine to bring these off-topic conversations to Discord (discord.gg/csharp) to discuss with the team and community there. Discord is much more free-form, where all sorts of conversations and topic flows are totally acceptable. CSharplang discussions need to stay on topic though. Thanks!

May I respond to your remarks about me?

Read the OP, it suggests adding another way, a new way to express a non terminating loop to the language, that actually _isn't_ trivial with C# (despite your claims to the contrary) and I think that mentioning this and the reasons for it is pertinent to the discussion, hence me saying what I said.

**Reply by HaloFour** [reply]
> Read the OP, it suggests adding another way, a new way to express a non terminating loop to the language, that actually isn't trivial with C# (despite your claims to the contrary) and I think that mentioning this and the reasons for it is pertinent to the discussion, hence me saying what I said.

Adding a `loop` keyword in this context would actually be pretty trivial since it's not possible for it to clash with existing syntax and it basically just duplicates an existing control statement.  There's nothing about the design of C# and how it parses grammar that would impact the implementation of this feature, it's entirely a question as to whether or not such a feature would be a good addition to the language.

**Reply by The-Futurist** [reply]
See: https://learn.microsoft.com/en-us/archive/blogs/ericlippert/asynchrony-in-c-5-part-six-whither-async

> We considered many heuristics that could make good guesses about whether a given "await" was intended as an identifier rather than a keyword, and did not like any of them.

That is to say, it **wasn't** trivial.

**Reply by HaloFour** [reply]
Loop wouldn't have an operand, that problem does not apply.

**Reply by CyrusNajmabadi** [reply]
> Read the OP, it suggests adding another way, a new way to express a non terminating loop to the language, that actually isn't trivial with C#

Your claim was " some of these suggestions would be trivial to deliver, no risk, little effort."

For this proposal, the issue is not risk/effort.  Nor is it that we have a C lineage that affects anything here.  The issue is *entirely* around if this is a *good* feature to introduce to the ecosystem, or if this actually would cause more issues (like confusion/bifurcation).

> that actually isn't trivial with C#

It really would be, especially as it could be defined as a trivial transformation to an existing construct (like `while(true)`).  As such, the lang design would be near nil, as would the impl.  But, again, that is not the concern here.  The concern is if this is a *good* change, not about if it's a trivial or not.

Note: this general point has come up in several discussions recently that you have participated in.  The discussions here do cover a wide gamut of topics, but those topics will always include: "is this a good thing for the language and ecosystem?" and also "should we do this?"  These questions are always going to be the core determiners of what happens with a discussion. 

Questions and discussions about complexity/triviality are certainly relevant.  But debating them isn't really sensical when the maintainers here (including myself) can weigh in with high precision as to what that would be.  For some topics it would be hard to give such an estimate (like NRT), but for things like `loop`, it's very easy to do so.

**Reply by CyrusNajmabadi** [reply]
> That is to say, it wasn't trivial.

You are conflating different topics.  `loop` is a different grammar construct than `await` (which is gramatically defined as `await expr`).  `loop` has no impl issues here.  But *again*, the lang-team itself is telling you that this is not relevant to this discussion.  The constraint here is whether this is a good idea or not for the *language* And *ecosystem*.  If so, we would do it.  Whether or not you think it's complex or not (which it isn't) doesn't have bearing on the discussion.

**Reply by CyrusNajmabadi** [reply]
@Korporal Based on several of your posts in this repo over the past few days (and in the past), i think you're operating under several misunderstandings and misapprehensions about how the language is designed (especially the grammar), and how it impacts potential future lang features.  I would *highly* recommend either creating a discussion on that topic, or coming to discord (discord.gg/csharp, `Roslyn` channel) to discuss this matter further.  We can help clarify this area, which will help many of your future conversations on this repo.

**Reply by The-Futurist** [reply]
> @Korporal Based on several of your posts in this repo over the past few days (and in the past), i think you're operating under several misunderstandings and misapprehensions about how the language is designed (especially the grammar), and how it impacts potential future lang features. I would _highly_ recommend either creating a discussion on that topic, or coming to discord (discord.gg/csharp, `Roslyn` channel) to discuss this matter further. We can help clarify this area, which will help many of your future conversations on this repo.

There really isn't anything to "clarify", admittedly in the case under discussion the grammar would support:

```cs
loop
{
}
```
and have no backward compatibility consequences from what I can see, so yes in this specific case it is easy, I was wrong. In general though adding new keywords does present challenges as Lippert's piece explains and as [this post too](https://blog.paranoidcoding.com/2022/04/11/lowercase-type-names.html) explains. 

So I have the clarity, thank you.

**Reply by CyrusNajmabadi** [reply]
> In general though adding new keywords does present challenges

IN general, changing any language can present challenges, depending on what the change is :)   All languages will have design choices that will impact what sorts of changes you can make, and what sorts of things are simple, what sorts of things are complex, and potentially what sorts of things are near impossible.

In c# though, it tends to be pretty simple teh vast majority of the time to make these sorts of grammar changes.  Occasionally, albeit *very rarely*, there are harder challenges.  Only twice though have we had to do anything complex to support these things.  And the second time we learned from the first, and took an approach that greatly reduced that complexity, making only marginally more difficult than any other sort of normal change here.

> explains and as [this post too](https://blog.paranoidcoding.com/2022/04/11/lowercase-type-names.html) explains.

These posts dive deep into these topics to share info for curious and enthusiastic readers about the rare cases that do arise.  They are not a conveyance about the general case.

**Reply by The-Futurist** [reply]
> > In general though adding new keywords does present challenges
> 
> IN general, changing any language can present challenges, depending on what the change is :) All languages will have design choices that will impact what sorts of changes you can make, and what sorts of things are simple, what sorts of things are complex, and potentially what sorts of things are near impossible.
> 
> In c# though, it tends to be pretty simple teh vast majority of the time to make these sorts of grammar changes. Occasionally, albeit _very rarely_, there are harder challenges. Only twice though have we had to do anything complex to support these things. And the second time we learned from the first, and took an approach that greatly reduced that complexity, making only marginally more difficult than any other sort of normal change here.
> 
> > explains and as [this post too](https://blog.paranoidcoding.com/2022/04/11/lowercase-type-names.html) explains.
> 
> These posts dive deep into these topics to share info for curious and enthusiastic readers about the rare cases that do arise. They are not a conveyance about the general case.

I agree with most of that, no argument but it is bordering on desperation to start warning developers to avoid lower case class names so that the language can more easily be changed:

>Hence rather than take one hit at a time when we release new C# language versions we took the broader view of allowing lower case type names is _hampering language evolution_.

That's a pretty general, or as he says "broad" view, so lets not minimize this, the grammar is crippled and this more or less proves that point, why on earth would one care what a user names their classes! this is not a good thing, not at all.

**Reply by CyrusNajmabadi** [reply]
@Korporal as before, general posts about your feelings around the c# grammar are *off topic* fort specific discussions.  

I have laid out what is allowed on csharplang.  This includes making a dedicated post on this topic (but keeping it on C#).  You are also welcome to discuss this matter on many of our other communities. Further discussions of the general grammar design and how you perceive it will not be permitted in specific discussions.  Thanks!

Specific posts on the grammar here are fine.  Going off onto tangents about the grammar in general are not.

**Reply by The-Futurist** [reply]
> @Korporal as before, general posts about your feelings around the c# grammar are _off topic_ fort specific discussions.
> 
> I have laid out what is allowed on csharplang. This includes making a dedicated post on this topic (but keeping it on C#). You are also welcome to discuss this matter on many of our other communities. Further discussions of the general grammar design and how you perceive it will not be permitted in specific discussions. Thanks!

Crazy, you reply to me and make debatable statements ("it tends to be pretty simple the vast majority of the time to make these sorts of grammar change") I respond to _you_ and you berate _me_ for discussing it here!! as you wish, I'm done.

**Comment by The-Futurist** [comment]
So, going back to the beginning of the thread (i.e. looping!), what could be the objection to adding 

```cs
forever
{

}
```

to the C# language? It's trivial to do, that's been established, is simpler than `for(;;)` or `while(true)` and surely better than:

```cs
repeat_what_I_just_did:
   do_stuff();
   do_more_stuff();
goto repeat_what_I_just_did;
```

(despite several recent recommendations to leverage `goto` more when we can) I mean what _is_ the downside here? really?

**Reply by 333fred** [reply]
I don't know that it's simpler than `while (true)`, to be entirely honest. It's a new concept to add to the language for a very niche scenario that nearly never occurs, for which we already have a simple, idiomatic, and completely understandable syntax form.

**Reply by spydacarnage** [reply]
> what _is_ the downside here? really?

It's still adding complexity to the language, however _technically_ trivial it is.

So it still has to pass the cost vs. benefit analysis...

**Reply by CyrusNajmabadi** [reply]
> is simpler than for(;;) or while(true) 

Here's the issue though. It's not really simpler. That's the power of idiom.  With each of the above, you are learning a single concept and applying it to a range of use cases. If we add a new form, that's something *new* to learn that only applies to a single use case. That's less simple. We don't want to clutter the language with single-use features used only for single cases. We want general features that scale across a gamut of important cases. 

If it was the case that the simple case was ill-served by the existing feature, we might feel differently.  But that isn't the case here.

**Reply by CyrusNajmabadi** [reply]
> despite several recent recommendations to leverage goto more when we can

The recommendation was to use goto for the loop break/continue case since you already need a label to demarcate the construct of interest.  Since you need both the label and the break/continue, it's similarly ergonomic to just have a label and a goto.  

Such an argument doesn't hold here as the ergonomics would be different (and it's slightly odd to bring up as no one has ever suggested that form here).

**Reply by The-Futurist** [reply]
> > is simpler than for(;;) or while(true)
> 
> Here's the issue though. It's not really simpler. That's the power of idiom. With each of the above, you are learning a single concept and applying it to a range of use cases. If we add a new form, that's something _new_ to learn that only applies to a single use case. That's less simple. We don't want to clutter the language with single-use features used only for single cases. We want general features that scale across a gamut of important cases.
> 
> If it was the case that the simple case was ill-served by the existing feature, we might feel differently. But that isn't the case here.

Yes, you do like to keep saying "idiom". If idioms where to be truly valued here we'd all be admiring and using COBOL with its maddening verbosity, a plethora of lovely idioms. It'll be interesting to hear your argument that "forever" is _not_ a simpler term for "forever" than "for (;;)" or "while (true)" are purported to be.

It's worth noting too that idioms and slang are often indistinguishable, in fact idioms are known for meaning something not easily inferred from the terms used to construct it, how a reduced ability to reason can be regarded as desirable in a formal language used across cultures to program hugely sophisticated systems, is, I must say, surprising to me, consider

> In [linguistics](https://en.wikipedia.org/wiki/Linguistics), idioms are usually presumed to be [figures of speech](https://en.wikipedia.org/wiki/Figures_of_speech) contradicting the [principle of compositionality](https://en.wikipedia.org/wiki/Principle_of_compositionality). That compositionality is the key notion for the analysis of idioms is emphasized in most accounts of idioms.[[6]](https://en.wikipedia.org/wiki/Idiom#cite_note-6)[[7]](https://en.wikipedia.org/wiki/Idiom#cite_note-7) This principle states that **the meaning of a whole should be constructed from the meanings of the parts that make up the whole.** In other words, one should be in a position to understand the whole if one understands the meanings of each of the parts that make up the whole.

See [idiom](https://en.wikipedia.org/wiki/Idiom) (emphasis mine).

If the principles you are espousing here (embracing of vulgarity), really do represent the strategy favored by your team, then I fear C# has an ugly future ahead.

**Reply by CyrusNajmabadi** [reply]
> Yes, you do like to keep saying "idiom".

Yes.  As both fred and i have mentioned, these forms are idiomatic and that is relevant to our decision making.  I'm not sure what else to tell you here.  You're welcome to think that such a think isn't relevant, or should be weighed differently.  Your feelings and opinions on this are certainly valid.  

Wrt to the term 'idiomatic', i'm using the following definition: 

> using, containing, or denoting expressions that are natural to a native speaker.

The forms you've mentioned above are completely natural to the C# community.  They are the trivial version of those constructs applied to the trivial case.  They are simple, understandable, and have broad acceptance.  You are totally free to believe that they are not sufficient and that a specialized construct be present to support this trivial case as well.  However, currently, no members of the C# language design group appears to agree with that view, and prefers the status quo where the existing constructs continue to be used in the same fashion as they have for 25 years.  Perhaps something will change here in the future.  But for now, that's the state of things.  I'm sorry if that's not palatable to you and your desires for the future of the language.

Finally, wrt to the rest of the post, please keep things civil and constructive.  The team has made no secret of the things it considers important, and these views and approaches have been what have directed the development of the language since v1.  If you have issue with that, then perhaps this isn't the best tool for you.  If you are interested in participating in the development of it though, then understanding how the team views these things is relevant and will continue to be part of the process, just as it always has been.  Thanks!

**Reply by CyrusNajmabadi** [reply]
@Korporal Further meta discussion about the design process around C# and the weights and balances the team gives to aspects of that design should be put into a different discussion.  This discussion should only be on the topic of 'loop' (or similar solutions to the concept of providing an alternative to `while (true)`).

**Reply by The-Futurist** [reply]
There are contradictions and epistemological flaws in your reasoning above, but I cannot disagree with you since doing so leads to such a post being labelled as "disruptive" or "off topic" or "meta discussion" and thus puts you in a position where your propositions cannot be challenged. Challenges can be neatly dismissed, hand waved away.

So I'll bid you farewell on this thread, I'm not willing to discuss a subject on such inequitable terms.

**Reply by HaloFour** [reply]
@Korporal 

You've been invited many times to join the Discord server if you want to have further discussions along those lines.  You've chosen to ignore that and to continue to argue here.

The terms are inequitable because you _are not a language design team member_.  This isn't a democracy.  Proposals aren't going to be accepted based on winning arbitrary logic games, nor are they going to be accepted based on your badgering over logic flaws.

C# loosely follows C philosophies which follow UNIX philosophy, which is to have simpler composable pieces that can be combined to solve larger problems.  Adding a `loop` or `forever` keyword that only satisfies a single and relatively uncommon use case directly contradicts those philosophies.  The fact that this use case can already be satisfied through the combination of existing language features and that these approaches are idiomatic by virtue that they have been the documented approach for doing so for the past 50 years in the C ecosystem makes it further unnecessary.  There is no friction in the language here, and trying to reduce from two tokens to one in a situation that might occur once, if ever, in any given program, is not a goal of the language.

**Comment by CyrusNajmabadi** [comment]
> I mean what is the downside here? really?

It's pretty much the same reasoning that came up with the `elsif` discussion.  A new way do to do something already easy and idiomatic.  It's practically no upsides and many downsides.
